tony-core/src/main/java/com/linkedin/tony/ReusablePort.java

```diff
@@ -44,54 +44,84 @@
  */
 final class ReusablePort extends ServerPort {
   private static final Log LOG = LogFactory.getLog(ReusablePort.class);
-  final EventLoopGroup eventLoopGroup;
-  final ChannelFuture future;
+  private final Process socketProcess;
+  private final int port;
+  public static final Path RESERVE_PORT_SCRIPT_PATH = requireNonNull(createPortReserveScript());
+  public static final String PORT_FILE_NAME_SUFFIX = "___PORT___";
+
+  private static Path createPortReserveScript() {
+    ClassLoader classloader = Thread.currentThread().getContextClassLoader();
+    final String reservePortScript = "reserve_reusable_port.py";
+    try {
+      // copy reserve_reusable_port.py from resource dir to a tmp dir
+      Path tempDir = Files.createTempDirectory("reserve_reusable_port");
+      tempDir.toFile().deleteOnExit();
+      try (InputStream stream = classloader.getResourceAsStream(reservePortScript)) {
+        Files.copy(stream, Paths.get(tempDir.toAbsolutePath().toString(), reservePortScript));
+      }
+      return Paths.get(tempDir.toAbsolutePath().toString(), reservePortScript);
+    } catch (IOException ex) {
+      return null;
+    }
+  }
 
-  ReusablePort(EventLoopGroup loopGroup, ChannelFuture future) {
-    this.eventLoopGroup = loopGroup;
-    this.future = future;
+
+  ReusablePort(Process socketProcess, int port) {
+    this.socketProcess = socketProcess;
+    this.port = port;
   }
 
-  private static void close(EventLoopGroup loopGroup, ChannelFuture future) {
-    if (future != null && future.channel().isOpen()) {
-      future.channel().close().awaitUninterruptibly();
+  private static void killSocketBindingProcess(Process process) {
+    Preconditions.checkNotNull(process);
```

Comment:
just some nits, I think you are mixing usage of Preconditions.checkNotNull and Objects.requireNonNull. I prefer always stick to one

Reply:
thanks, updated.